test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -93,46 +88,62 @@ test_ucp_tag::request* test_ucp_tag::request_alloc()
     return req;
 }
 
-void test_ucp_tag::send_callback(void *request, ucs_status_t status)
+void test_ucp_tag::send_callback(void *request, ucs_status_t status,
+                                 void *user_data)
 {
-    struct request *req = (struct request *)request;
+    struct request *req = (struct request *)(user_data ?: request);
     ucs_assert(req->completed == false);
     req->status    = status;
     req->completed = true;
 }
 
+/* TODO: deprecated, remove after complete migration to new API */
+void test_ucp_tag::send_callback(void *request, ucs_status_t status)
+{
+    send_callback(request, status, NULL);
+}
+
 void test_ucp_tag::recv_callback(void *request, ucs_status_t status,
-                                 ucp_tag_recv_info_t *info)
+                                 const ucp_tag_recv_info_t *info,
+                                 void *user_data)
 {
-    struct request *req = (struct request *)request;
+    struct request *req = (struct request *)(user_data ?: request);
     ucs_assert(req->completed == false);
     req->status    = status;
     req->completed = true;
     if (status == UCS_OK) {
-        req->info      = *info;
+        req->info  = *info;
     }
 }
 
-void test_ucp_tag::wait(request *req, int buf_index)
+/* TODO: deprecated, remove after complete migration to new API */
+void test_ucp_tag::recv_callback(void *request, ucs_status_t status,
+                                 ucp_tag_recv_info_t *info)
+{
+    recv_callback(request, status, info, NULL);
+}
+
+void test_ucp_tag::wait(void *ucx_req, void *user_data, int buf_index)
 {
+    request *req     = (request*)(user_data ?: ucx_req);
```

Comment:
we should always wait on ucx_req, not user_data, since ucx_req is the one returned from send/recv operation